### DISCUSSION TITLE
How to get anchor of an URL

### DISCUSSION THREAD

**Discussion by Romaixn** [vercel/next.js#53337] [Help] [answered]
### Summary

How to retrieve the anchor of a page (http://localhost/page#anchor, for example)? Is it possible to obtain it using `next/navigation`? I tried with `document.URL`, but I got an error saying `document is undefined`.

**Comment by Nadeera3784** [comment, marked as answer]
In Next.js, you can access the current URL and its anchor (also known as a fragment) using the `window.location` object, but there is a caveat. The `window` object is only available on the client-side, and Next.js server-side rendering (SSR) does not have access to it. That's why you encountered the "document is undefined" error when trying to use `document.URL` in Next.js.

To properly handle this situation, you can check if you are running in a client-side environment before accessing the `window.location` object. You can do this by using the `typeof window` check. Here's how you can retrieve the anchor of the current page in a Next.js component:

```jsx
import { useEffect } from 'react';

const PageComponent = () => {
  useEffect(() => {
    // Check if we are in a client-side environment
    if (typeof window !== 'undefined') {
      // Access the anchor using window.location.hash
      const anchor = window.location.hash;
      console.log('Anchor:', anchor);
    }
  }, []);

  return (
    <div>
      {/* Your page content */}
    </div>
  );
};

export default PageComponent;
```

By doing this, you ensure that the code accessing `window.location` only runs on the client-side after the component is mounted, preventing any server-side errors.

For example, if you access `http://localhost/page#anchor`, the `console.log` statement should output: "Anchor: #anchor". This means you successfully retrieved the anchor from the URL.

Remember that when using client-side navigation in Next.js (e.g., using `next/link` or `next/router`), the page will not reload entirely, and the component might not be re-mounted, so the useEffect will not be triggered again unless you explicitly reload the page. This means if you navigate to another page within the same Next.js app and it has a different anchor, the useEffect in the new page component will not fire automatically. If you need to capture changes in the anchor on client-side navigation, you can use Next.js router events to listen for changes and update your component accordingly.
